lms/djangoapps/branding_by_referer/configuration_by_referer.py

```diff
@@ -0,0 +1,161 @@
+"""
+Overwites some site configuration according to the referer
+"""
+import copy
+import json
+
+from django.utils.deprecation import MiddlewareMixin
+from django.utils.six import iteritems
+from django.utils.six.moves.urllib.parse import urlparse
+from openedx.core.djangoapps.site_configuration import helpers as configuration_helpers
+from openedx.core.djangoapps.user_api.models import UserPreference
+
+
+class SetConfigurationByReferer(MiddlewareMixin):
+    """
+    This middleware checks if the current logged-in user has a marketing site referer preference.
+    If so, it overwrites some values of the site configuration based on the referer.
+    It depends on the SetBrandingByReferer middleware to apply the configuration overwrites, as that
+    middleware sets the referer preference for the current logged-in user. Therefore this middleware
+    must be located below the SetBrandingByReferer one in the MIDDLEWARE_CLASSES setting
+    """
+    MARKETING_SITE_REFERER = 'MARKETING_SITE_REFERER'
+    OVERRIDE_MKTG_REFERER_KEY = 'CURRENT_OVERRIDE_REFERER'
+    CONFIGURATION_BY_REFERER_KEY = 'CONFIGURATION_BY_REFERER'
+
+    def process_request(self, request):
+        """
+        Process request middleware method.
+        """
+        self.validate_reset_configuration(request)
+
+        if not self.is_feature_enabled():
+            return None
+
+        referer_domain_preference = self.get_stored_referer_preference(request)
+        if not referer_domain_preference:
+            return None
+
+        # just return if the site configuration is not enabled
+        if not configuration_helpers.is_site_configuration_enabled():
+            return None
+
+        # Save a copy of the original SiteConguration dict
+        request.original_site_conf = copy.deepcopy(
+            configuration_helpers.get_current_site_configuration().values
+        )
+
+        self._update_conf(request, referer_domain_preference)
+
+    def _override_marketing_urls(self, referer_domain):
```

Comment:
Why do you make it so specific to the marketing urls? would it not be easier to ask the configuration for the referere to have the exact configuration you want for the marketing urls and override using that without making the process know much about what is it overriding?

Reply:
My concern was defining the mktg urls dict for each referer in the site configuration and make it even bigger and difficult to manipulate manually (is it a concern for you?). So I forced this to happen for all referers, by replacing the global MKTG url domains with the referer domain, as LLPA does not want to have links to the wrong marketing site.

Reply:
It is not much of a concern for me. Basically the object is as complicated as can be already. But at least it is straight forward. If we open up the door of having special methods for each key we might overcomplicate the code and it will never be enough